rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -400,79 +418,112 @@ impl CanisterQueues {
+            let msg = pop_and_advance(input_queue, &mut self.pool);
+
             // If the input queue is non-empty, re-enqueue the sender at the back of the
             // input schedule queue.
             if input_queue.len() != 0 {
-                input_schedule.push_back(sender);
+                self.input_schedule.reschedule(*sender, input_queue_type);
+            } else {
+                self.input_schedule.pop(input_queue_type).unwrap();
             }
 
-            self.input_queues_stats -= InputQueuesStats::stats_delta(QueueOp::Pop, &msg);
-            self.memory_usage_stats -= MemoryUsageStats::stats_delta(QueueOp::Pop, &msg);
-            debug_assert!(self.stats_ok());
-
-            return Some(msg.into());
+            if let Some(msg) = msg {
+                debug_assert_eq!(Ok(()), self.test_invariants());
+                debug_assert_eq!(Ok(()), self.schedules_ok(&|_| InputQueueType::RemoteSubnet));
+                return Some(msg.into());
+            }
         }
 
+        debug_assert_eq!(Ok(()), self.test_invariants());
+        debug_assert_eq!(Ok(()), self.schedules_ok(&|_| InputQueueType::RemoteSubnet));
         None
     }
 
     /// Peeks the next canister input queue message.
-    fn peek_canister_input(&self, input_queue: InputQueueType) -> Option<CanisterMessage> {
-        let input_schedule = match input_queue {
-            InputQueueType::LocalSubnet => &self.local_subnet_input_schedule,
-            InputQueueType::RemoteSubnet => &self.remote_subnet_input_schedule,
-        };
-        if let Some(sender) = input_schedule.front() {
-            // The sender's input queue.
-            let input_queue = &self.canister_queues.get(sender).unwrap().0;
-            let msg = input_queue.peek().unwrap();
-            return Some(msg.clone().into());
+    ///
+    /// It is possible for an input schedule to contain an empty input queue if e.g.
+    /// all messages in said queue have expired / were shed since it was scheduled.
+    /// Requires a `&mut self` reference to be able advance to the first non-empty
+    /// queue in the input schedule, to achieve amortized `O(1)` time complexity.
+    fn peek_canister_input(&mut self, input_queue_type: InputQueueType) -> Option<CanisterMessage> {
+        while let Some(sender) = self.input_schedule.peek(input_queue_type) {
+            if let Some(item) = self
+                .canister_queues
+                .get(sender)
+                .and_then(|(input_queue, _)| input_queue.peek())
+            {
+                let msg = self
+                    .pool
+                    .get(item.id())
+                    .expect("stale reference at the head of input queue");
+                debug_assert_eq!(Ok(()), self.test_invariants());
+                debug_assert_eq!(Ok(()), self.schedules_ok(&|_| InputQueueType::RemoteSubnet));
+                return Some(msg.clone().into());
+            }
+
+            // Queue was garbage collected or is empty.
```

Comment:
Should we comment here and potentially in other places that peek succeeded above and hence the unwrap can not fail?